### PR TITLE
[MGDOBR-403] Improve deployment scripts for local dev

### PR DESCRIPTION
**Pull request by radtriste**
[MGDOBR-403](https://issues.redhat.com/browse/MGDOBR-403) 

See also https://github.com/radtriste/sandbox/blob/deploy_script/dev/README.md

```bash
$ ./dev/bin/deploy.sh -h
Usage: deploy.sh [OPTIONS]

Options:
  -l                  Use images built locally (they will be built automatically)
  -s                  Start the cluster
  -a                  Deploy all resources on the cluster

Examples:
  # Simple deployment. Only dev image, aka ingress and executor will be deployed to cluster
  sh deploy.sh

  # Start Kind cluster and deploy everything on it with local build of images
  sh deploy.sh -a -l -s
```

```bash
$ ./dev/bin/undeploy.sh -h
Usage: undeploy.sh [OPTIONS]

Options:
  -s                  Stop the cluster

Examples:
  # Just deploy all resources from cluster
  sh undeploy.sh

  # Undeploy and stop cluster
  sh undeploy.sh -s
```

_Running integration-tests locally would be done in 2 commands:_
```
$ ./dev/bin/deploy.sh -l -c minikube -s -a
$ ./integration-tests/run-local-tests.sh
```

**Detailed modifications:**
- [x] support for Kind cluster
- [x] created `dev/config` folder and moved there `dev/bin/localconfig` and `dev/bin/environment`, that latter one becoming `interactionconfig` (`dev/bin/common.sh` contains automatic migration commands)
- [x] `dev/_deploy` contains all information relative to the local deployment
  - `dev/_deploy/kustomize` contains a copy of `kustomize` folder with base and  `minikube` overlay. All link/modifications are then done in this folder
  - `dev/_deploy/local_env` contains information about the current local deployment (cluster, cluster_ip, manager_url, keycloak_url, interaction_ob_token if any)
- [x] `dev/bin/configure.sh` renamed to `dev/bin/common.sh`
  - added some `configure_*` methods. scripts don't source with configuration profiles, we call directly the `config`
  - contains useful dirs
  - source in all other script files
  - other methods to read/write to `dev/deploy/local_env` file
  - `get_keycloak_access_token` refactored to handle getting token from internal keycloak url. Useful with Kind as localhost is not accessible from pods (can only reference it external with `localhost`, no external IP ... )
  - `get_sso_access_token` added to retrieve token from Red Hat SSO
- [x] added `deploy.sh` and `undeploy.sh` scripts
- [x] renamed other `dev/bin/*` scripts for more coherence
- [x] refactored a bit the `dev/bin/interaction` scripts to reuse methods from `dev/bin/common.sh` scripts and avoid to get a token for each call. Configuration is now done via `dev/config/interaction-config` file
- [x] updated docs

Please make sure that your PR meets the following requirements:

- [ ] Your code is properly formatted according to [this configuration](https://github.com/kiegroup/kogito-runtimes/tree/main/kogito-build/kogito-ide-config)
- [ ] Your commit messages are clear and reference the JIRA issue e.g: "[MGDOBR-1] - $clear_explanation_of_what_you_did"
- [ ] All new functionality is tested
- [ ] Pull Request title is properly formatted: `MGDOBR-XYZ Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains link to any dependent or related Pull Request
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket

<details>
<summary>
How to trigger pipelines and use the bots:
</summary>

* <b>Run the end to end pipeline</b>  
  Annotate the pull request with the label: `safe to test`. If you want to run the pipeline again, remove and add it again.

* <b>Rebase the pull request</b>  
  Comment with: `/rebase`.

</details>

### CODE REVIEW

File: dev/README.md
Comment (kostola) [inline dev/README.md:45]: you should change the name of the directory also in the link destination, otherwise it will be inconsistent (`[dev/config](config)`)

File: dev/bin/configure.sh
Comment (kostola) [inline dev/bin/configure.sh:136]: I think we should uniform the function names to snake case instead of mixing it with camel case

File: dev/bin/configure.sh
Comment (kostola) [inline dev/bin/configure.sh:55]: Why are we exporting all these variables?

**Review by kostola — CHANGES_REQUESTED** [review]

File: dev/bin/configure.sh
Comment (kostola) [reply dev/bin/configure.sh:55]: Also, we should use lowercase names for local variables and uppercase only for environment variables to avoid conflicts (see https://stackoverflow.com/a/673940)

File: dev/README.md
Comment (radtriste) [reply dev/README.md:45]: I need to update the whole doc with this new script :)

File: dev/bin/configure.sh
Comment (radtriste) [reply dev/bin/configure.sh:136]: you are right. gonna update those methods

File: dev/bin/configure.sh
Comment (radtriste) [reply dev/bin/configure.sh:55]: those are meant to be used in other scripts. so they should be considered as env vars. that is why we are exporting them

Comment (kostola) [reply dev/bin/configure.sh:55]: But all other scripts are sourcing `configure.sh`, so these variables will be already set without exporting them (which increases the risk of conflicts with existing env variables).

I would use uppercase variables only for actual env variables that the user can use to configure the scripts (e.g. `MANAGED_KAFKA_INSTANCE_NAME`).

File: .github/actions/kind/action.yml
Comment (danielezonca) [inline .github/actions/kind/action.yml:21]: Why this change? I would prefer to keep the version fixed to reduce CI instability

**Review by danielezonca — COMMENTED** [review]
Just one minor comment, I'm not doing a proper review.
My only consideration is that it looks like our scripting "toolkit" is increasing fast so we might consider to cover them with some tests and/or move to a different language/framework that is easier to test (i.e. jbang/picocli but there are many of them).
@robpblake 
Wdyt?

File: .github/actions/kind/action.yml
Comment (radtriste) [reply .github/actions/kind/action.yml:21]: > Just one minor comment, I'm not doing a proper review. My only consideration is that it looks like our scripting "toolkit" is increasing fast so we might consider to cover them with some tests and/or move to a different language/framework that is easier to test (i.e. jbang/picocli but there are many of them). @robpblake Wdyt?

I do see your point. Right now I just aggregated what is currently existing in an "orchestration" script. Not much new.
I guess we could later migrate to jbang or other

**Comment by radtriste** [comment]
> Just one minor comment, I'm not doing a proper review. My only consideration is that it looks like our scripting "toolkit" is increasing fast so we might consider to cover them with some tests and/or move to a different language/framework that is easier to test (i.e. jbang/picocli but there are many of them). @robpblake Wdyt?

I do see your point. Right now I just aggregated what is currently existing in an "orchestration" script. Not much new.
I guess we could later migrate to jbang or other

**Comment by radtriste** [comment]
for testing the shell script, we can use the [bats testing](https://opensource.com/article/19/2/testing-bash-bats) 
I can add some tests to this PR

File: dev/bin/configure.sh
Comment (radtriste) [reply dev/bin/configure.sh]: done

File: .github/actions/kind/action.yml
Comment (radtriste) [reply .github/actions/kind/action.yml]: fallback done

File: dev/bin/configure.sh
Comment (radtriste) [reply dev/bin/configure.sh]: I updated all of those. we now use only lowercase names for variables coming from `configure.sh` (renamed to `common.sh`)

File: dev/README.md
Comment (radtriste) [reply dev/README.md]: doc updated

File: dev/bin/common.sh
Comment (danielezonca) [inline dev/bin/common.sh:120]: In many scripts I see the usage of `configure_manager_url`. Is it the same? (and btw, where is defined?)

File: dev/README.md
Comment (danielezonca) [inline dev/README.md:326]: To be updated? I don't think the script is named `manager-run.sh` anymore (same comment for the other links)

File: dev/bin/common.sh
Comment (danielezonca) [inline dev/bin/common.sh:87]: This should be the same defined in the .github/actions/kind/action.yml 
Can you add a comment to keep them aligned?

File: dev/bin/common.sh
Comment (danielezonca) [inline dev/bin/common.sh:98]: Why `1.21.10` while minikube has `1.21.6`?

File: dev/bin/cluster-stop.sh
Comment (danielezonca) [inline dev/bin/cluster-stop.sh:24]: This is not strictly related with this PR but I think we should decide if we test kind or minikube for local development and align CI, current kind + minikube scripting situation has an overhead that we can probably get rid of.

@robpblake 
Wdyt?

File: dev/bin/cluster-start.sh
Comment (danielezonca) [inline dev/bin/cluster-start.sh:11]: To be updated?

File: dev/bin/cluster-stop.sh
Comment (danielezonca) [inline dev/bin/cluster-stop.sh:11]: To be updated

File: dev/bin/cluster-kind-stop.sh
Comment (danielezonca) [inline dev/bin/cluster-kind-stop.sh:4]: To be updated

File: dev/bin/undeploy.sh
Comment (danielezonca) [inline dev/bin/undeploy.sh:50]: Yes I agree with this change, current grep based on `ob-` prefix is error prone: that prefix is configurable

File: dev/README.md
Comment (radtriste) [reply dev/README.md:11]: gonna add it

Comment (radtriste) [reply dev/README.md:11]: don't know how much time my PR will take to be merged. so better to merge the one from @lucamolteni first and will rebase mine then

File: dev/bin/localconfig
Comment (radtriste) [reply dev/bin/localconfig:14]: oops. will remove

File: dev/bin/local-manager-stop.sh
Comment (radtriste) [reply dev/bin/local-manager-stop.sh:15]: I see what you mean. gonna merge into a `local-service-stop.sh` file

File: dev/bin/interaction/get-all-bridges.sh
Comment (radtriste) [reply dev/bin/interaction/get-all-bridges.sh:8]: gonna verify that

File: dev/bin/interaction/create-processor.sh
Comment (radtriste) [reply dev/bin/interaction/create-processor.sh:3]: could the problem of rhoas CLI be that it targets automatically the console.redhat.com URL and not the local one ?

File: dev/bin/common.sh
Comment (radtriste) [reply dev/bin/common.sh:120]: gonna check

File: dev/bin/common.sh
Comment (radtriste) [reply dev/bin/common.sh:87]: I was more thinking about an `.env` file in the root of the project. wdyt ?

Comment (danielezonca) [reply dev/bin/common.sh]: Even better :)

Comment (radtriste) [reply dev/bin/common.sh]: implemented. please have a look. I had to update the `kind` github action

File: dev/README.md
Comment (lucamolteni) [reply dev/README.md]: Yes rebase will fix this

**Comment by lucamolteni** [comment]
if we decide to go for this, I think we should communicate it to the team properly as the structure of the scripts seems to have changed significantly

**Comment by radtriste** [comment]
> if we decide to go for this, I think we should communicate it to the team properly as the structure of the scripts seems to have changed significantly

Yes that was my intention :)
once merged, I will write on #dev room

**Comment by manstis** [comment]
> once merged, I will write on #dev room

I prefer emails to be honest, pings on Google Chat are easily lost (or it's just me).

File: dev/bin/local-manager-stop.sh
Comment (radtriste) [reply dev/bin/local-manager-stop.sh:10]: @danielezonca I updated the stop scripts. Can you please have a look ?

Comment (radtriste) [reply dev/bin/local-manager-stop.sh]: it will now look to the `target/*-dev.jar` file and do the kill based on that file.

File: dev/bin/logs-shard.sh
Comment (danielezonca) [inline dev/bin/logs-shard.sh:18]: Please convert to kubectl or add oc to required tools

File: dev/bin/interaction/create-processor.sh
Comment (danielezonca) [reply dev/bin/interaction/create-processor.sh:3]: No currently we don't have CLI support for OB in rhoas but I think we should do it ~soon so that we can use it directly

File: dev/bin/logs-shard.sh
Comment (radtriste) [reply dev/bin/logs-shard.sh:18]: woops ...

**Comment by radtriste** [comment]
- rebased
- migration should be easier
  - added migration guide in dev README file
  - config file migration will be done automatically

@lucamolteni @manstis @danielezonca @sutaakar please have a look

File: dev/README.md
Comment (sutaakar) [inline dev/README.md:24]:
```suggestion
Configuration can be done via environment variables or via the `dev/config/localconfig` file.
```

File: dev/README.md
Comment (sutaakar) [inline dev/README.md:47]:
```suggestion
#### Minikube configuration for macOS users
```

File: dev/README.md
Comment (sutaakar) [inline dev/README.md:45]: What should this refer to?

File: dev/README.md
Comment (sutaakar) [inline dev/README.md:170]:
```suggestion
### Setup Managed Kafka
```

File: dev/README.md
Comment (radtriste) [reply dev/README.md:45]: I just forgot to remove the TODO ...

```suggestion
```

File: dev/README.md
Comment (radtriste) [inline dev/README.md:46]:
```suggestion
```

File: dev/README.md
Comment (radtriste) [reply dev/README.md:120]: true... forgot to remove it after the rebase

Comment (radtriste) [reply dev/README.md:120]:
```suggestion
```

**Review by r00ta — COMMENTED** [review]
That's very good job @radtriste , but I have some concerns about the maintenance of those scripts since you are going to leave the project. 
Personally, I'm already struggling when I have to update the scripts for small changes because they require some knowledge, and if we put more and more overhead in the scripts it is going to be even worse. 

Don't get me wrong, I know there is a lot of effort in this but I'm not sure this is the right thing to do now.

**Comment by radtriste** [comment]
Honestly, I just reused what is existing (with a few renaming) and added a "flow" script to run one after the other, as described in the documentation.
I did not add any magic... The complexity did not changed.

But if you feel like it should not be merged, then don't. I am also ok with that.

File: dev/bin/cluster-stop.sh
Comment (robpblake) [reply dev/bin/cluster-stop.sh:19]: @danielezonca Yes agreed. A single approach with either KIND or Minikube makes the most sense. I don't have a strong opinion either way - both should work just fine - but we don't have capacity to support both.

Comment (jomarko) [reply dev/bin/cluster-stop.sh:19]: +1 to point, the amount of introduced scripts increases the maintain costs. Assuming @radtriste is moving to other project after this is finished. I vote for having single option.

File: dev/README.md
Comment (jomarko) [inline dev/README.md:361]: What is the recommended way for this migration. In past , I have never run `bin/common.sh` on its own. Is it the same? If not, we should explicitly mention, that it is first step of migration.

File: dev/bin/interaction/create-processor.sh
Comment (jomarko) [reply dev/bin/interaction/create-processor.sh:3]: @danielezonca is such task already in jira or available to work on? I mean the rhoas CLI commands for smart events.

**Review by jomarko — COMMENTED** [review]
Thank you for the PR @radtriste

File: dev/README.md
Comment (radtriste) [reply dev/README.md:361]: this script is a utils script which is called from all other scripts. You should not need to run it.
Here is maybe a better sentence. Let me know if you are ok with that @jomarko 

```suggestion
On the execution of any script, it should call the `bin/common.sh` script, which will automatically transfer the configuration files to the `config` folder.  
Here are the moves done:
```

Comment (jomarko) [reply dev/README.md:361]: that sounds better to me @radtriste , yes

**Pull request closed without merging**
